test/test_common/global.cc

```diff
@@ -0,0 +1,70 @@
+#include "test/test_common/global.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Test {
+
+Globals& Globals::instance() {
+  static Globals* h = new Globals;
+  return *h;
+}
+
+std::string Globals::describeActiveSingletonsHelper() {
+  std::string ret;
+  Thread::ReleasableLockGuard map_lock(map_mutex_);
+  for (auto& p : singleton_map_) {
+    std::unique_ptr<Singleton>& singleton = p.second;
+    ASSERT(singleton != nullptr);
+    if (singleton->ptr_ != nullptr) {
+      absl::StrAppend(&ret, "Unexpected active singleton: ", p.first, "\n");
+    }
+  }
+  return ret;
+}
+
+Globals::Singleton& Globals::get(const std::string& type_name, MakeObjectFn make_object) {
+  Thread::ReleasableLockGuard map_lock(map_mutex_);
+  std::unique_ptr<Singleton>& singleton = singleton_map_[type_name];
+
+  if (singleton == nullptr) {
+    // The first time constructing this object, we'll be constructing the
+    // Singleton for the first time, populating it with a fresh instance,
+    // so no need to take the singleton's mutex. But we do need to hold
+    // the map mutex as we are installing the singleton object in the
+    // singleton_map.
+    singleton = std::make_unique<Singleton>(make_object());
+    map_lock.release();
+    return *singleton;
+  }
+
+  // Subsequent accesses will must lock the singleton's mutex to safely populate
+  // .second. However, we can relinquish map_mutex_ as we are no longer
+  // modifying the map; just the .second of the MutexSingleton pointed to by the
+  // map.
+  Thread::LockGuard singleton_lock(singleton->mutex_);
+  if (singleton->ptr_ == nullptr) {
+    ASSERT(singleton->ref_count_ == 0);
+    singleton->ptr_ = make_object();
+  }
+  ++singleton->ref_count_;
```

Comment:
How come you're not using a shared_ptr/weak_ptr pattern here but doing it by hand?

Reply:
I thought about this for a bit; couldn't figure out how to make the pieces fit given the heterogenous map and casting.

Reply:
What about `std::shared_ptr<Singleton>` (and corresponding weak_ptr), with a `void *` still inside the object to provide the type cast abilities?